Docs: Fundamentals of Block Development - File structure of a block (closed) (#56316)
WordPress/gutenberg: docs/getting-started/fundamentals-block-development/file-structure-of-a-block.md

# File structure of a block

> At its simplest, a block in the WordPress block editor is a JSON object with a specific set of properties that is registered in the system so it can be used.

Reviewer: This is incorrect. At its simplest it is a registerBlockType function call with some parameters passed in JS

Block.json is just an abstraction that makes this easier and allows us to more easily share the configuration between JS and PHP.

Author: Thanks a lot for this feedback @fabiankaegy 🙌
I have adjusted the sentences in this page to take this into account.

> The recommended way to register a block is via a plugin. With the `create-block` tool you can quickly scaffold the structure of the files required to create a plugin tha registers a block. 

Reviewer 2: ```suggestion
The recommended way to register a block is via a plugin. With the `create-block` tool you can quickly scaffold the structure of the files required to create a plugin that registers a block. 
```

### `<plugin-file>.php`

A block is usually added to the block editor using a WordPress plugin. In the main PHP file of the plugin the block is usually registered on the server side.

<div class="callout callout-info">
For more on creating a WordPress plugin see <a href="https://developer.wordpress.org/plugins/plugin-basics/">Plugin Basics</a>, and <a href="https://developer.wordpress.org/plugins/plugin-basics/header-requirements/">Plugin Header requirements</a> for explanation and additional fields you can include in your plugin header.
</div>

### `package.json`

[`package.json`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json) is a configuration file for a NodeJs project. In this file you define the NPM dependencies of the block and the scripts used for local work.

### `src` folder

In a standard project you'll place your block files in the `src` folder. By default, the build process with `wp-scripts `will take files from this folder and will generate the bundled files in the `build` folder. 

### `block.json`

> This files contains the metadata of the block, and it's required for its registration. Among other things it provides properties to define the files you need to be part of the definition of the block:

Reviewer 2: I found this difficult to read. Does this describe block.json or the list of files below the text? I think I struggled because of the plural form.

Reviewer: Again this file is not required at all :). 

It just makes things easier.

Author: I have updated this part taking into account the  feedback of both you 🙏

Reviewer 2: Isn't block.json required to be able to style the block with theme.json? If only I could recall exactly where I got that information from. I assume it is related to API versions...

Reviewer: @carolinan It may be the fact that a block needs to be registered on the server for global styles for the block to get printed correctly. 

But you can register a block on the server without using the `block.json` file. 

The file just makes it easier :)

- `editorScript` → `index.js` by default
- `style` → `style-index.css` (output build for `style.scss`) by default
- `editorStyle` → `index.css` (output build for `editor.scss`) by default
- `render` → `render.php` by default
- `viewScript` → `view.js` by default

### `index.js`

The `index.js` file (or any other file defined in the `editorScript` property of `block.json`) is the entry point file for javascript files and it usually registers the block in the client. In a standard structure it imports the `edit.js` and `save.js` files to get functions required in block registration.

### `edit.js`

The `edit.js` file usually returns the `edit` function used in block registration that defines the React component that will be loaded on the Editor when the block is loaded or inserted.

### `save.js`

The `save.js` file usually returns the `save` function used in block registration that defines the markup used to save the block in the DB.

### `style.scss`

> The `style.scss` file contains the styles of the block that will be loaded in both the editor and the frontend. In the build process this file is converted into `style-index.css` which is usually defined at `style` property in `block.json`

Reviewer 2: Will it be clear enough that this is a scss / sass file, or do we need to add something more about the file format and the processing?

Author: I have added an info box providing some more context about these `.scss` files 👍

### `editor.scss`

The `editor.scss` file contains the additional styles applied to the block only in the editor’s context. In the build process this file is converted into `index.css` which is usually defined at `editorStyle` property in `block.json`

### `render.php`

The `render.php` file (or any other file defined in the `render` property of `block.json`) defines the server side process that returns the markup for the block when there is a request from the frontend. If this file is defined, it will take precedence over any other ways to render the block's markup for the frontend.

### `view.js`

> The `view.js` file (or any other file defined in the `view` property of `block.json`) will be loaded in the front-end when the block is displayed.

Reviewer: In the viewScript property

# File structure of a block

> The most common way to create a custom block is via a JSON object that is registered in the system so it can be used.

Reviewer: ```suggestion
The most common way to register a custom block is via a JSON file that contains all of the metadata of the block.
```

> The recommended way to register a block is via a plugin. With the `create-block` tool you can quickly scaffold the structure of the files required to create a plugin that registers a block. 

Reviewer: ```suggestion
It is recommended to register blocks within plugins to ensure they stay available when a theme gets switched. With the `create-block` tool you can quickly scaffold the structure of the files required to create a plugin that registers a block. 
```

### `<plugin-file>.php`

A block is usually added to the block editor using a WordPress plugin. In the main PHP file of the plugin the block is usually registered on the server side.

<div class="callout callout-info">
For more on creating a WordPress plugin see <a href="https://developer.wordpress.org/plugins/plugin-basics/">Plugin Basics</a>, and <a href="https://developer.wordpress.org/plugins/plugin-basics/header-requirements/">Plugin Header requirements</a> for explanation and additional fields you can include in your plugin header.
</div>

### `package.json`

[`package.json`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json) is a configuration file for a NodeJs project. In this file you define the NPM dependencies of the block and the scripts used for local work.

### `src` folder

In a standard project you'll place your block files in the `src` folder. By default, the build process with `wp-scripts `will take files from this folder and will generate the bundled files in the `build` folder. 

### `block.json`

> This file contains the metadata of the block, and it's used to simplify the definition and registration of the block. 

Reviewer: ```suggestion
This file contains the metadata of the block, and it's used to simplify the definition and registration of the block both in the browser and on the server. 
```

> This file contains the metadata of the block, and it's used to simplify the definition and registration of the block both in the browser and on the server. 

Reviewer: ```suggestion
This file contains the metadata of the block, and it's used to simplify the definition and registration of the block both in the client and on the server. 
```

I made this mistake here. I much prefer how you used `client` later on in the document :)

The most relevant properties that can be defined in a `block.json` to set the files involved in the block's behaviour, output or style are:
- The `editorScript` property, usually set with the path of a bundled `index.js` file (output build from `src/index.js`).
- The `style` property, usually set with the path of a bundled `style-index.css` file (output build from `src/style.scss`).
- The `editorStyle` property, usually set with the path of a bundled `index.css` (output build from `src/editor.scss`).
- The `render` property, usually set with the path of a bundled `render.php` (output copied from `src/render.php`).
- The `viewScript` property, usually set with the path of a bundled `view.js` (output copied from `src/view.php`).

[![Open Build Output Diagram in excalidraw](https://developer.wordpress.org/files/2023/11/file-structure-build-output.png)](https://excalidraw.com/#json=c22LROgcG4JkD-7SkuE-N,rQW_ViJBq0Yk3qhCgqD6zQ "Open Build Output Diagram in excalidraw")

### `index.js`

> The `index.js` file (or any other file defined in the `editorScript` property of `block.json`) is the entry point file for javascript files and it usually registers the block in the client. In a standard structure it imports the `edit.js` and `save.js` files to get functions required in block registration.

Reviewer: ```suggestion
The `index.js` file (or any other file defined in the `editorScript` property of `block.json`) is the entry point file for javascript that should only get loaded in the editor. It is responsible for calling the `registerBlockType` function to register the block on the client. In a standard structure it imports the `edit.js` and `save.js` files to get functions required in block registration.
```

### `edit.js`

> The `edit.js` file usually returns the `edit` function used in block registration that defines the React component that will be loaded on the Editor when the block is loaded or inserted.

Reviewer: ```suggestion
The `edit.js` commonly gets used to contain the React component that gets used in the editor for our block. It usually exports a single component that then gets passed to the `edit` property of the `registerBlockType` function in the `index.js` file.
```

### `save.js`

> The `save.js` file usually returns the `save` function used in block registration that defines the markup used to save the block in the DB.

Reviewer: ```suggestion
The `save.js` is similar to the `edit.js` file in that it exports a single React component. This component generates the static HTML markup that gets saved to the Database. 
```

### `style.scss`

> The `style.scss` file contains the styles of the block that will be loaded in both the editor and the frontend. In the build process this file is converted into `style-index.css` which is usually defined at `style` property in `block.json`

Reviewer: I think it is a little missleading to only talk about SCSS here. WP scripts also supports regular straight up CSS and postCSS here.

Author: Thanks for this feedback @fabiankaegy  👌
I have expanded the explanation to include more CSS formats accepted by `wp-scripts`. 
Let me know what you think
